docs: vTPM spec (#6802)
vatesfr/xen-orchestra: docs/vtpm.md

```ts
 vTPM Object
 {

>   id: string,

Reviewer: `uuid`.

  is_protected: boolean, // Currently unused by the XAPI, may be used in future.

>   is_unique: boolean // Currently unused by the XAPI, may be used in future.

Reviewer: Remove for now if unused.

 }
```

- `is_protected`: The TPM is encrypted whenever it's transferred or at stored to disk (and exported)
- `is_unique`: The TPM contents are exclusive to that VM, that means it has never been cloned (and maybe never exported nor imported)

> vTPM features are only enabled if XAPI is recent enough (check for version >= 8.3).

Reviewer: Alternative checks:
- `pool.restrictions.restrict_vtpm === 'false'`

- vTPM
  - At VM creation
    - A toggle will be displayed in advanced settings under `boot firmware`.
    - The default value is `false`, except if `template.platform.vtpm` is `"true"`.
      - A message will be displayed below the toggle if:

>         - The value is `true`: _Please note that once the vTPM is used, its deletion is strongly discouraged as it will lead to data loss._

Reviewer: Not necessary, a warning confirmation modal at vTPM creation will be enough.

>         - The value is `false` and the `template.platform.vtpm` is `"true"`: _The VM cannot be created without vTPM_

Reviewer: ```suggestion
        - The value is `false` and the `template.platform.vtpm` is `"true"`: _This template requires a vTPM, if you proceed, the VM will likely not be able to boot_
```

```ts
 vTPM Object
 {
  id: string,
  is_protected: boolean, // Currently unused by the XAPI, may be used in future.
  is_unique: boolean // Currently unused by the XAPI, may be used in future.
 }
```

- `is_protected`: The TPM is encrypted whenever it's transferred or at stored to disk (and exported)
- `is_unique`: The TPM contents are exclusive to that VM, that means it has never been cloned (and maybe never exported nor imported)

> vTPM features are only enabled if XAPI is recent enough (check for `pool.restrict` or version (>= 8.3)).

Reviewer 2: Exact check to be decided. I think XCP-ng/XS version >= 8.3 should be enough as a check. Otherwise I can provide more precise criteria.

- vTPM
  - At VM creation

>     - A toggle will be displayed in advanced settings under `boot firmware`.

Reviewer 2: The exact label of the toggle could be specified. Something like "Attach a vTPM to the VM", or just "vTPM"? We also probably need to define a tooltip for this label - or other means through which you provide explanations to users in XO - to explain what it is.

Author: I think `vTPM` is enough. With `Attach a vTPM` it looks like we can attach an already created vTPM to the VM.

Otherwise, what about: `enable vTPM`?

>     - The default value is `false`, except for Windows >= W11.

Reviewer 2: The exact check is not on the version of the OS, but on the contents of the template. We now know what exactly defines the need for vTPM in a template. See https://github.com/xenserver/guest-templates-json/blob/master/json/windows-11.json.

Extract:
```
    "platform": {
        "acpi_laptop_slate": "1",
        "vtpm": "true" <------- THIS
    },
```

      - A message will be displayed below the toggle if:
        - The value is `true`: _Please note that once the vTPM is used, its deletion is strongly discouraged as it will lead to data loss._

>         - The value is `false` and the VM is a Windows 11 or later: _VMs under Windows 11 or later cannot be created without vTPM_

Reviewer 2: This must be based on the template defining the need for a vTPM or not, see above comment.

    - The toggle is disabled if XAPI is not recent enough or if `boot firmware` is not `UEFI`.
      - If disabled, a tooltip will be displayed:

>         - If XAPI is not recent enought: _The vTPM support is only available on pool XCP-ng/XS 8.3 or later._

Reviewer 2: ```suggestion
        - If XAPI is not recent enough: _vTPM support is only available on pools running XCP-ng/XS 8.3 or later._
```

>         - If `boot firmware` is not `UEFI`: _To use the vTPM, the boot firmware need to be UEFI_

Reviewer 2: ```suggestion
        - If `boot firmware` is not `UEFI`: _A UEFI boot firmware is necessary to use a vTPM_
```

  - VM advanced tab

>     - A new section _VTPM_ will be created just under `Xen settings`.

Reviewer 2: ```suggestion
    - A new section _vTPM_ will be created just under `Xen settings`.
```

    - If no vTPM is attached to the current VM:
      - A button labeled _Create a vTPM_ will be displayed.
      - The button is disabled if XAPI is not recent enough or if `boot firmware` is not `UEFI`.

>         - If disabled, a tooltip will be displayed:

Reviewer 2: Could a tooltip also be displayed when enabled, to explain what a vTPM is?

Author: I think it would be better to have an information icon (`i`), which would redirect you to documentation explaining what vTPM is.
This icon could be displayed just after the label of the `vTPM` section and should also be present on the VM creation page.

Reviewer 2: This works, but means we'll need the documentation to exist. Let's not do it for now and add it later.

>           - If XAPI is not recent enought: _The vTPM support is only available on pool XCP-ng/XS 8.3 or later._

Reviewer 2: ```suggestion
          - If XAPI is not recent enough: _vTPM support is only available on pools running XCP-ng/XS 8.3 or later._
```

>           - If `boot firmware` is not `UEFI`: _To use the vTPM, the boot firmware need to be UEFI_

Reviewer 2: ```suggestion
          - If `boot firmware` is not `UEFI`: __A UEFI boot firmware is necessary to use a vTPM__
```

>       - The button will trigger a confirmation modal which warn the user: _Please note that once the vTPM is used, its deletion is strongly discouraged as it will lead to data loss._

Reviewer 2: I first thought this was a mistake and was a copy and paste error, but I realize you really intend on warning users at creation time. I'm not sure this is necessary (same in VM create page). There already is a big warning when they click the Delete button.

    - If vTPM is attached to the current VM:
      - A button labeled _Delete the vTPM_ will be displayed.
      - The button will trigger a confirmation modal wich warn the user: _If the vTPM is in use, removing it will result in a dangerous data loss. This action is irreversible. Are you sure you want to remove the vTPM?_

>       - Below the button, the following information will be displayed:

Reviewer 2: If it doesn't break XO UI guidelines, I'd put the vTPM information above the Delete button, rather than below.

        - vTPM ID: sliced ID (as we do for backup ID or others, only 4 characters will be displayed and a copy to clipboard button will be available)
        - Is unique: boolean (if used)

>         - Is protected: boolean (if used)

Reviewer 2: To be decided: whether to display these properties or not. I think we could start with not displaying them, and start displaying them later, if they begin actually meaning something.
